kernel/kernel-api/src/main/java/io/delta/kernel/internal/rowtracking/RowTracking.java

```diff
@@ -135,6 +127,131 @@ public static Optional<DomainMetadata> createNewHighWaterMarkIfNeeded(
         : Optional.empty();
   }
 
+  /**
+   * Reassigns baseRowIds and defaultRowCommitVersions for AddFile actions during conflict
+   * resolution when row tracking is supported.
+   *
+   * <p>This method should handle two scenarios:
+   *
+   * <ol>
+   *   <li>Reassigning fields previously set by this transaction that need updating
+   *   <li>Assigning missing fields when a concurrent winning transaction added row tracking support
+   * </ol>
+   *
+   * <p>Currently, only the first scenario is handled. The second scenario cannot occur in Kernel
+   * since protocol changes between losing and winning transactions are not permitted.
+   *
+   * @param snapshot the snapshot of the table that this transaction is reading at
+   * @param winningRowIdHighWatermark the latest row ID high watermark from the winning transactions
+   * @param attemptCommitVersion the commit version of this losing transaction
+   * @param lastWinningVersion the last winning commit version
+   * @param dataActions the data actions that this losing transaction was trying to commit
+   * @return an iterable of data actions with reassigned baseRowIds and defaultRowCommitVersions
+   */
+  public static CloseableIterable<Row> reassignBaseRowIdAndDefaultRowCommitVersion(
```

Comment:
It might make sense to merge the logic to reassign row tracking information + update domain metadata with the logic to assign them initially during commit.
These are essentially the same except that:
- Initial assignment: only assign and increment high-water mark if no baseRowId / default row commit versions 
- Reassignment: only reassign and increment high-water mark if baseRowId > previous high-water mark, default commit version = previous commit version

You could achieve that by passing an optional previous high water mark / commit version: None during initial commit, set for reassignment during conflict resolution